File: 剑指offer/40.最小的k个数.py
from typing import List
import heapq


class Solution:
    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        if k == 0:
            return []

        heap = [-x for x in arr[:k]]

        heapq.heapify(heap)

        for i in range(k, len(arr)):
            if arr[i] < -heap[0]:
                heapq.heappop(heap)
                heapq.heappush(heap, -arr[i])

        return [-x for x in heap]

File: 剑指offer/40.最小的k个数.cpp
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> getLeastNumbers(vector<int>& arr, int k)
    {
        if (k == 0) {
            return {};
        }

        priority_queue<int, vector<int>> pq;

        for (int i = 0; i < k; i++) {
            pq.push(arr[i]);
        }

        for (int i = k; i < arr.size(); i++) {
            if (arr[i] < pq.top()) {
                pq.pop();
                pq.push(arr[i]);
            }
        }
        vector<int> ans;

        for (int i = 0; i < k; i++) {
            ans.push_back(pq.top());
            pq.pop();
        }

        return ans;
    }
};